common/decompressingWriter.go

```diff
@@ -0,0 +1,138 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"io"
+	"time"
+)
+
+type decompressingWriter struct {
+	pipeWriter  *io.PipeWriter
+	workerError chan error
+}
+
+const decompressingWriterCopyBufferSize = 256 * 1024 // 1/4 the size that we usually write to disk with (elsewhere in codebase). 1/4 to try to keep mem usage a bit lower, without going so small as to compromize perf
+var decompressingWriterBufferPool = NewMultiSizeSlicePool(decompressingWriterCopyBufferSize)
+
+var ECompressionType = CompressionType(0)
+
+type CompressionType uint8
+
+func (CompressionType) None() CompressionType        { return CompressionType(0) }
+func (CompressionType) ZLib() CompressionType        { return CompressionType(1) }
+func (CompressionType) GZip() CompressionType        { return CompressionType(2) }
+func (CompressionType) Unsupported() CompressionType { return CompressionType(255) }
+
+// NewDecompressingWriter returns a WriteCloser which decompresses the data
+// that is written to it, before passing the decompressed data on to a final destination
+func NewDecompressingWriter(destination io.WriteCloser, ct CompressionType) io.WriteCloser {
+	preader, pwriter := io.Pipe()
+
+	d := &decompressingWriter{
+		pipeWriter:  pwriter,
+		workerError: make(chan error, 1),
+	}
+
+	// start the output processor worker
+	go d.worker(ct, preader, destination, d.workerError)
+
+	return d
+}
+
+func (d decompressingWriter) decompressorFactory(tp CompressionType, preader *io.PipeReader) (io.ReadCloser, error) {
+	switch tp {
+	case ECompressionType.ZLib():
+		return zlib.NewReader(preader)
+	case ECompressionType.GZip():
+		return gzip.NewReader(preader)
+	default:
+		return nil, errors.New("unexpected compression type")
+	}
+}
+
+func (d decompressingWriter) worker(tp CompressionType, preader *io.PipeReader, destination io.WriteCloser, workerError chan error) {
+
+	defer func() {
+		_ = destination.Close() // always close the destination file before we exit, since its a WriteCloser
+		_ = preader.Close()
+	}()
+
+	// make the decompressor. Must be in the worker method because,
+	// like the rest of read, this reads from the pipe.
+	// (Factory reads from pipe to read the zip/gzip file header)
+	dec, err := d.decompressorFactory(tp, preader)
+	if err != nil {
+		workerError <- err
+		return
+	}
+
+	// Now read from the pipe, decompressing as we go, until
+	// reach EOF on the pipe (or encounter an error)
+	b := decompressingWriterBufferPool.RentSlice(decompressingWriterCopyBufferSize)
+	_, err = io.CopyBuffer(destination, dec, b) // returns err==nil if hits EOF, as per docs
```

Comment:
Consider adding a comment mentioning that this would cause the goroutine to hold until the EOF or an error is hit

Reply:
That comment already exists, above.